### PR TITLE
mds: path and uid/gid cap enforcement

### PR DESCRIPTION
**Pull request by liewegas**
- path restrictions are complete
- uid/gid restrictions are incomplete, but opt-in -- they only
kick in if he mds cap has a uid specified; it should be safe
to merge as is.

There is a change in the MClientCaps message here to add
the uid/gid so we should be sure we want it before merge.

### CODE REVIEW

**Comment by gregsfortytwo** [comment]
@liewegas, I'll see if I can review this from just the diff, but I thought you were going to squash down all that meandering about trying to find the right interface...we usually ask for a much neater patch series than this. :)

**Comment by gregsfortytwo** [comment]
test_auth_caps is targeted at a vstart cluster, which is a bit awkward for our teuthology runs targeting all workunits. (that is, it will fail and trigger reporting in teuthology.) :(
Also it's assuming the default setup will be to mount as client.admin; again valid for vstart but maybe not elsewhere. It would probably be best to transcribe that into a CephFSTestCase in ceph-qa-suite? Or maybe we could leave the actual file code in shell but move the setup stuff so it can run in either situation; not sure how to do that best though...

File: qa/workunits/fs/test_path_caps.sh [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+unlink mnt.foo/foo/r
+rmdir mnt.foo/foo/p
+
+# everything else is not
+expect_false mkdir mnt.foo/bar/x
+expect_false mkdir mnt.foo/food
+expect_false ln -s symlink mnt.foo/x
+expect_false ln -s symlink mnt.foo/food
+expect_false unlink mnt.foo/bar/s
+expect_false rmdir mnt.foo/bar/q
```
Comment (gregsfortytwo) [inline qa/workunits/fs/test_path_caps.sh]: I'm a bit confused, this test verifies we can mount the root despite not being allowed to read its path. Is that what we want to do? Do we expect to see any of the root directory's contents, or only the stuff in foo? Can we add some tests which actually cover that?

File: src/Makefile.am [hunk 2]
```diff
@@ -149,7 +149,7 @@ endif
 core-daemons: ceph-mon ceph-osd ceph-mds radosgw
 admin-tools: monmaptool osdmaptool crushtool ceph-authtool
 base: core-daemons admin-tools \
-	cephfs ceph-syn ceph-conf \
+	cephfs ceph-fuse ceph-syn ceph-conf \
```
Comment (gregsfortytwo) [inline src/Makefile.am]: I can't tell what this "base" target is for, but I wonder if we also have to fix something in CMake.

File: src/client/Client.cc [hunk 3]
```diff
@@ -4872,7 +4890,7 @@ int Client::mount(const std::string &mount_root)
       break;
   }
 
-  assert(root_ancestor->is_root());
+  //assert(root_ancestor->is_root());
```
Comment (gregsfortytwo) [inline src/client/Client.cc]: No commented-out code, please.

File: src/client/Client.cc [hunk 4]
```diff
@@ -5592,11 +5614,29 @@ int Client::_setattr(Inode *in, struct stat *attr, int mask, int uid, int gid,
...
   // make the change locally?
+  if ((in->cap_dirtier_uid >= 0 && uid != in->cap_dirtier_uid) ||
+      (in->cap_dirtier_gid >= 0 && gid != in->cap_dirtier_gid)) {
+    ldout(cct, 10) << __func__ << " caller " << uid << ":" << gid
+		   << " != cap dirtier " << in->cap_dirtier_uid << ":"
+		   << in->cap_dirtier_gid << ", forcing sync setattr"
+		   << dendl;
+    if (!mask)
+      mask |= CEPH_SETATTR_CTIME;
+    goto force_request;
```
Comment (gregsfortytwo) [inline src/client/Client.cc]: Can you talk through how this check and the switch to a synchronous server call keeps things safe?

**Comment by liewegas** [comment]
@gregsfortytwo rebased some of this away, and then added a pile of fixes on top that came up when rewriting all of the tests in ceph_test_libcephfs.  I think the uid/gid stuff is in decent shape now, but I mostly just want to get the path enforcement in place for infernalis.

Also fixed the OpenLayout test (which was making assumptions about data pool name).

**Comment by gregsfortytwo** [comment]
@liewegas how's the testing on this branch?

**Comment by liewegas** [comment]
i'm waiting for a clean master so i can rebase and re-run.. just let me 
knwo when

**Comment by gregsfortytwo** [comment]
I merged a bunch of branches yesterday that have stuff pretty clean — last
greg-fs-testing run had 3 failures on monitor valgrind issues and 1 failure
to populate /sys/fuse. Hopefully it won't take more than a couple days to
test and merge in several more fixes (including one for that) but it's a
lot happier than it was.

**Comment by liewegas** [comment]
https://github.com/ceph/ceph-qa-suite/pull/586 seems to fix it!  http://pulpito.ceph.com/sage-2015-09-28_09:48:01-fs-wip-sage-testing---basic-multi/

**Pull request closed without merging**